enhancements/substitutes-for.md

```diff
@@ -0,0 +1,386 @@
+
+The implications of this for opm add + the catalog db are that the substitutes-for metadata must be stored in the DB associated w/ each bundle.
+
+The expected flow is:
+
+1) opm add inserts the new bundle following all existing logic, but also adding the substitutes-for metadata
+2) Upon a list-bundles request, the operator-registry should return a substitution-corrected list of bundles based on substitutes replaces/skips references.
+3) In addition, the registry export command will need to perform substitutions on the CSVs as it exports them, to ensure their replaces/skips references are correct.
+
+
+Complications:
+
+1) the CSV itself contains other version references(beyond skips/replaces), but it should be ok not to perform updates on them because those references are not used by anything making upgrade decisions
+
+2) skipRange logic could theoretically get invalidated (1.2 substitutes for 1.1, but some bundle's skipRange is "skipRange: <1.2").  Since this feature is primarily being added for CVE rebuilds which we expect to only bump the patch release, this should simply be documented as a known limitation/caveat around using substitutes-for (that it does not allow substitution consideration during skipRange upgrades)
+
+3) If C substitutes B and B substituted A, it should be assumed that C substitutes A.  The db metadata and substitution processing needs to reflect this.
+
+
+## Motivation
+
+The primary motivation of this feature is to provide a way for the tooling to automatically rebuild an 
+operator bundle in response to CVEs in the operator's dependencies and then add the new bundle to the 
+operator catalog in such a way that clusters can upgrade to the new bundle from the unpatched bundle, 
+as well as upgrade from the patched bundle to newer bundles.
+
+### Goals
+
+* Patched bundles can be published that result in an intact upgrade graph, without having to modify
+other previously published bundles. (e.g. users do not end up stranded because they upgrade to a rebuilt
+bundle which cannot be upgraded to the head of another channel).
+* Future bundles that reference the unpatched version of a bundle, when a patched version exists in the
+catalog, have their references updated to include the patched versions.
+* Clusters running a version of the bundle that predates the vulnerable bundle are not forced to
+upgrade into the vulnerable bundle before being able to upgrade to the patched bundle.
+* The code changes need to be contained within the catalog and opm binary so that it does not require
+OCP cluster changes.  It must work across existing 4.5+ versions of OCP clusters.
+
+### Non-Goals
+
+* Fully refactoring the upgrade graph management system out of bundle metadata so that upgrade graphs
+can be modified independently of bundle modification/publication.  (This is a future goal for OLM, however)
+
+
+## Proposal
+
+The proposal is to add a new field (tentatively named "substitutes-for") to CSVs, the value of which is one or
```

Comment:
It would be nice if we could make this use `properties` explicitly (see shelved PR: https://github.com/operator-framework/enhancements/pull/35/files) instead of adding more and more things the CSV that then need to be extracted into properties.

Reply:
not sure i agree.  a properties hash is not an easily discoverable api (what keys are valid, what do they mean?) and it's more error prone to program to.

What is the benefit of putting this in properties vs having an explicit field for it?  (is properties a CSV field today?)

Reply:
+1, I think using `properties` leaves us more aligned with where we want to go. 

@ecordell if we go that route, I think this enhancement would need to define a new property/constraint type. WDYT?

Reply:
I think I'll write up something separate for `properties` that I think will clarify why I was suggesting it.

A CSV field should be fine for now, since we don't have another way to explicitly set them anyway. But I would suggest that `substitutes-for` be a field like `skips` that is read as input to opm but not on-cluster / present in the schema.